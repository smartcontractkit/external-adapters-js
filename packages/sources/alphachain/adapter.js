const { Requester, Validator } = require('@chainlink/external-adapter')
const { util } = require('@chainlink/ea-bootstrap')

const customError = (data) => {
  if (data.status !== '200') return true
  return false
}

const customParams = {
  base: ['base', 'from', 'coin'],
  quote: ['quote', 'to', 'market'],
}

const execute = (input, callback) => {
  const validator = new Validator(input, customParams)
  if (validator.error) return callback(validator.error.statusCode, validator.errored)

  const jobRunID = validator.validated.id
  const url = 'https://alpha-chain2.p.rapidapi.com/data-query'
  const host = 'alpha-chain2.p.rapidapi.com'
  const headers = {
    'content-type': 'application/octet-stream',
    'x-rapidapi-host': host,
    'x-rapidapi-key': util.getRandomRequiredEnv('API_KEY'),
    useQueryString: true,
  }
  const base = validator.validated.data.base.toUpperCase()
  const quote = validator.validated.data.quote.toUpperCase()

  const params = {
    from_symbol: base,
    to_symbol: quote,
    chainlink_node: true,
  }

  const config = {
    url,
    params,
    headers,
  }

  Requester.request(config, customError)
    .then((response) => {
      response.data.result = Requester.validateResultNumber(response.data, ['result'])
      callback(response.status, Requester.success(jobRunID, response))
    })
    .catch((error) => callback(500, Requester.errored(jobRunID, error)))
}

module.exports.execute = execute
