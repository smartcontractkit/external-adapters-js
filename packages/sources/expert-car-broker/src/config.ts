import { Requester } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

export const NAME = 'EXPERT_CAR_BROKER'

export const DEFAULT_ENDPOINT = 'feed'
export const DEFAULT_BASE_URL = 'https://prices.expertcarbroker.workers.dev/'

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix, true)
  config.api.baseURL = config.api.baseURL || DEFAULT_BASE_URL
  return config
}
