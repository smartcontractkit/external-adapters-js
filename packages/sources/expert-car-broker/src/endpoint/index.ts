export * as feed from './feed'
