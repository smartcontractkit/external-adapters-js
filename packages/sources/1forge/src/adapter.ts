import { Builder, Logger, Requester, Validator } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
  MakeWSHandler,
} from '@chainlink/ea-bootstrap'
import { DEFAULT_WS_API_ENDPOINT, makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const makeWSHandler = (config?: DefaultConfig): MakeWSHandler => {
  const getSubscription = (symbol?: string, subscribe = true): string => {
    if (!symbol) return ''
    return subscribe ? `subscribe_to|${symbol}` : `unsubscribe_from|${symbol}`
  }
  const getSymbol = (input: AdapterRequest) => {
    const validator = new Validator(
      input,
      endpoints.quotes.inputParameters,
      {},
      { shouldThrowError: false },
    )
    if (validator.error) return
    if (Array.isArray(validator.validated.data.quote)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.quote} supplied as quote. Only non-array tickers can be used for WS`,
      )
      return
    }
    if (Array.isArray(validator.validated.data.base)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.base} supplied as base. Only non-array tickers can be used for WS`,
      )
      return
    }
    const symbol = validator.validated.data.base.toUpperCase()
    const convert = validator.validated.data.quote.toUpperCase()
    return `${symbol}/${convert}`
  }
  const parseResponse = (response: string) => {
    const [type, data] = response.split('|')
    return { type, data: JSON.parse(data) }
  }

  return () => {
    const defaultConfig = config || makeConfig()

    return {
      connection: {
        url: DEFAULT_WS_API_ENDPOINT,
      },
      subscribe: (input) => getSubscription(getSymbol(input)),
      unsubscribe: (input) => getSubscription(getSymbol(input), false),
      subsFromMessage: (message: any) => {
        if (!message.data || message.data.indexOf('update') === -1) return
        const { data } = parseResponse(message.data)
        return getSubscription(data.s)
      },
      isError: () => false,
      filter: (message: any) => message.data.indexOf('update') !== -1,
      toResponse: (message: any) => {
        const { data } = parseResponse(message.data)
        return Requester.success('1', { data: { result: data.p } })
      },
      onConnectChain: [
        {
          payload: `login|${defaultConfig.apiKey}`,
          filter: (message: any) => message.data.indexOf('post_login_success') !== -1,
        },
      ],
    }
  }
}
