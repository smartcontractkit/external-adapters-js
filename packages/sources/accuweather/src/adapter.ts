import { Builder } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
} from '@chainlink/ea-bootstrap'
import { makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
