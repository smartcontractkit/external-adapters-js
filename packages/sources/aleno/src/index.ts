import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
