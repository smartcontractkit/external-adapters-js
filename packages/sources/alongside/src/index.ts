import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
