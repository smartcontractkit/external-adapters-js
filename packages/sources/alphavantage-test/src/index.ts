import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { forex } from './endpoint'

export const adapter = new PriceAdapter({
  defaultEndpoint: forex.name,
  name: 'ALPHAVANTAGE',
  config,
  endpoints: [forex],
  rateLimiting: {
    tiers: {
      free: {
        rateLimit1m: 5,
        rateLimit1h: 20.83,
      },
      '49.99': {
        rateLimit1m: 75,
      },
      '99.99': {
        rateLimit1m: 150,
      },
      '149.99': {
        rateLimit1m: 300,
      },
      '199.99': {
        rateLimit1m: 600,
      },
      '249.99': {
        rateLimit1m: 1200,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
