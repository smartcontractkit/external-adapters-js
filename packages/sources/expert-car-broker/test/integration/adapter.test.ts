import { Requester } from '@chainlink/ea-bootstrap'
import { assertError, assertSuccess } from '@chainlink/ea-test-helpers'
import { AdapterRequest } from '@chainlink/types'
import { makeExecute } from '../../src/adapter'

describe('execute', () => {
  const jobID = '1'
  const execute = makeExecute()

  describe('successful calls @integration', () => {
    const requests = [
      {
        name: 'id not supplied',
        testData: { data: { product: 'ferrari-f12tdf', feedId: 1 } },
      },
      {
        name: 'F12 TDF feed 1',
        testData: { id: jobID, data: { product: 'ferrari-f12tdf', feedId: 1 } },
      },
      {
        name: 'F12 TDF feed 2',
        testData: { id: jobID, data: { product: 'ferrari-f12tdf', feedId: 2 } },
      },
      {
        name: 'F12 TDF feed 3',
        testData: { id: jobID, data: { product: 'ferrari-f12tdf', feedId: 3 } },
      },
    ]

    requests.forEach((req) => {
      it(`${req.name}`, async () => {
        const data = await execute(req.testData as AdapterRequest)
        assertSuccess({ expected: 200, actual: data.statusCode }, data, jobID)
        expect(data.result).toBeGreaterThan(0)
        expect(data.data.result).toBeGreaterThan(0)
      })
    })
  })

  describe('error calls @integration', () => {
    const requests = [
      {
        name: 'unknown product',
        testData: { id: jobID, data: { product: 'NOT_A_REAL_PRODUCT', feedId: 1 } },
      },
      {
        name: 'unknown feedId',
        testData: { id: jobID, data: { product: 'ferrari-f12tdf', feedId: 999999 } },
      },
    ]

    requests.forEach((req) => {
      it(`${req.name}`, async () => {
        try {
          await execute(req.testData as AdapterRequest)
        } catch (error) {
          const errorResp = Requester.errored(jobID, error)
          assertError({ expected: 500, actual: errorResp.statusCode }, errorResp, jobID)
        }
      })
    })
  })
})
