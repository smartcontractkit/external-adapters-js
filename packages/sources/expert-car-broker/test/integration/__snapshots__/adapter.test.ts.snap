// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute feed api should return success 1`] = `
{
  "data": {
    "result": 482421,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 482421,
  "statusCode": 200,
}
`;
